### PR TITLE
chore: allow to connect using the new sidebar COMPASS-7710

### PR DESCRIPTION
**Pull request by kmruiz**
### Checklist
- [ ] New tests and/or benchmarks are included
- [ ] Documentation is changed or added
- [ ] I have signed the MongoDB Contributor License Agreement (https://www.mongodb.com/legal/contributor-agreement)

## Motivation and Context

- [ ] Bugfix
- [ ] New feature
- [ ] Dependency update
- [ ] Misc

## Types of changes

- [ ] *Backport Needed*
- [ ] Patch (non-breaking change which fixes an issue)
- [ ] Minor (non-breaking change which adds functionality)
- [ ] Major (fix or feature that would cause existing functionality to change)

### CODE REVIEW

**Comment by alenakhineika** [comment]
I have noticed that bootstrap fails in your branch when downloading fonts, when in main, no such issue.

```
FetchError: request to https://cloud.mongodb.com/static/font/EuclidCircularA-Semibold-WebXL.woff2 failed, reason: getaddrinfo ENOTFOUND cloud.mongodb.comDKpuz/5A9Fh4CEpyLScKR4Vcx8PqZAOQWapHQFcG4Z93jnNvMUUR00ZbC/YEgKSGkn+lcw== Completed in 56ms
    at ClientRequest.<anonymous> (/Users/alena.khineika/www/compass/node_modules/minipass-fetch/lib/index.js:97:14)
    at ClientRequest.emit (node:events:519:28)
    at TLSSocket.socketErrorListener (node:_http_client:500:9)
    at TLSSocket.emit (node:events:531:35)
    at emitErrorNT (node:internal/streams/destroy:169:8)
    at emitErrorCloseNT (node:internal/streams/destroy:128:3)
    at process.processTicksAndRejections (node:internal/process/task_queues:82:21) {
  code: 'ENOTFOUND',
  errno: 'ENOTFOUND',
  syscall: 'getaddrinfo',
  hostname: 'cloud.mongodb.com',
  type: 'system'
}
```

**Comment by kmruiz** [comment]
> I have noticed that bootstrap fails in your branch when downloading fonts, when in main, no such issue.

I didn't see the issue neither in CI or in my local machine, it seems like a temporary network issue in your laptop, maybe a VPN or something. Can you try again?

The URL seems correct and was unchanged (https://cloud.mongodb.com/...)  so it doesn't seem related to this PR.

File: packages/compass-connections/src/hooks/use-connection-status.ts
Comment (gribnoysup) [inline packages/compass-connections/src/hooks/use-connection-status.ts:16]: It's not allowed to use hooks dynamically

File: packages/compass-sidebar/src/components/multiple-connections/saved-connections/saved-connection.tsx
Comment (gribnoysup) [inline packages/compass-sidebar/src/components/multiple-connections/saved-connections/saved-connection.tsx:62]: There is a leafygreen warning icon, why are we making a copy?

File: packages/compass-sidebar/src/components/multiple-connections/sidebar.tsx
Comment (gribnoysup) [inline packages/compass-sidebar/src/components/multiple-connections/sidebar.tsx:78]: Once again, just so we continue to be on the same page: this code doesn't belong in the sidebar and should be refactored out to a connection plugin at some point. If you are still planning to use only COMPASS-7397 as an overarching ticket for this work, it would be good to add a todo here

File: packages/compass-sidebar/src/components/multiple-connections/sidebar.tsx
Comment (gribnoysup) [inline packages/compass-sidebar/src/components/multiple-connections/sidebar.tsx:85]: This probably should have an id based on a connection info id, otherwise one connection closes toasts for other connections. As the process is async and is not even visually blocked by the modal of any kind, this is very confusing: you can see progress or errors for multiple ongoing connections

File: packages/compass-sidebar/src/components/multiple-connections/sidebar.tsx
Comment (gribnoysup) [inline packages/compass-sidebar/src/components/multiple-connections/sidebar.tsx:125]: Very unexpected that error toast has a timeout, that's not a pattern we use anywhere else in the app

File: packages/compass-components/src/components/with-status-marker.tsx
Comment (gribnoysup) [inline packages/compass-components/src/components/with-status-marker.tsx:16]: We usually don't put components in compass-components that are very specific and are only used by a single place in the whole application

File: packages/compass-connections/src/hooks/use-connection-status.ts
Comment (gribnoysup) [inline packages/compass-connections/src/hooks/use-connection-status.ts:11]: This interface seems too broad: it requires us to get the whole connection info before being able to access the status while the only required value here is an id, can we limit this to actually required parameters?

File: packages/compass-sidebar/src/components/multiple-connections/sidebar.tsx
Comment (gribnoysup) [inline packages/compass-sidebar/src/components/multiple-connections/sidebar.tsx:203]: You probably can remove this comment now that you implemented it?

File: packages/compass-sidebar/src/components/multiple-connections/sidebar.tsx
Comment (gribnoysup) [inline packages/compass-sidebar/src/components/multiple-connections/sidebar.tsx:148]: It doesn't seem to be used by anything in the `useConnections`, why are we adding this?

File: packages/compass-components/src/components/with-status-marker.tsx
Comment (kmruiz) [reply packages/compass-components/src/components/with-status-marker.tsx:41]: Screenshot 2024-03-20 at 12 21 07

It's slightly different, it's green and has a background, as it overlaps with an icon. Look how the current spinner looks:

Screenshot 2024-03-20 at 12 24 50

File: packages/compass-components/src/components/with-status-marker.tsx
Comment (kmruiz) [reply packages/compass-components/src/components/with-status-marker.tsx:101]: Didn't realise it was a bit different, thanks! I'll fix it.

File: packages/compass-connections/src/hooks/use-connection-status.ts
Comment (kmruiz) [reply packages/compass-connections/src/hooks/use-connection-status.ts:16]: Ah right, I always forget about that rule.

I'll make the parameter mandatory.

File: packages/compass-sidebar/src/components/multiple-connections/sidebar.tsx
Comment (kmruiz) [reply packages/compass-sidebar/src/components/multiple-connections/sidebar.tsx:78]: I think I'll use this ticket for this change, as it should be small, instead of using COMPASS-7397.

File: packages/compass-sidebar/src/components/multiple-connections/sidebar.tsx
Comment (kmruiz) [reply packages/compass-sidebar/src/components/multiple-connections/sidebar.tsx:125]: Yeah Maurizio also was a bit surprised on this behaviour, but it's in the design. If the pattern is that error toasts are persistent, I think it's safe to remove the timeout and force the user to close it.

File: packages/compass-components/src/components/with-status-marker.tsx
Comment (kmruiz) [reply packages/compass-components/src/components/with-status-marker.tsx:16]: Well it's going to be used in different places in the sidebar, and it's a relatively small and simple, but I can move it to the sidebar.

File: packages/compass-connections/src/hooks/use-connection-status.ts
Comment (kmruiz) [reply packages/compass-connections/src/hooks/use-connection-status.ts:11]: Is there any reason to do so? I think it's more safe to just:

```
const connection = useConnectionInfo()
const status = useConnectionStatus(connection)
```

than

```
const connection = useConnectionInfo()
const status = useConnectionStatus(connection.id)
```

as requiring just the id is just an implementation detail of useConnectionStatus.

File: packages/compass-sidebar/src/components/multiple-connections/sidebar.tsx
Comment (kmruiz) [reply packages/compass-sidebar/src/components/multiple-connections/sidebar.tsx:148]: Ah, it was used before and probably kept it after the merge with Himanshu's PR and forgot to remove it. Thanks for pointing out!

File: packages/compass-connections/src/hooks/use-connection-status.ts
Comment (gribnoysup) [reply packages/compass-connections/src/hooks/use-connection-status.ts:11]: Can you elaborate what does it mean for id to be an implementation detail in this case? How else other parts of the UI should be able to communicate that they are attached to a certain connection?

The reason to do so is that we don't want to make a requirement for some parts of the application to depend on the whole connection info object (that they generally don't care about), potentially causing unneccessary subscriptions to the state updates, heavy deep check to bypass referencial equality issues, and complex state propagation scenarios in cases where passing an id that uniquely identifies a connection should be enough.

File: packages/compass-sidebar/src/components/multiple-connections/saved-connections/saved-connection.tsx
Comment (kmruiz) [reply packages/compass-sidebar/src/components/multiple-connections/saved-connections/saved-connection.tsx:62]: Didn't realise. Fixing it.

File: packages/compass-components/src/components/with-status-marker.tsx
Comment (gribnoysup) [reply packages/compass-components/src/components/with-status-marker.tsx:16]: Small and simple is not a deciding factor for putting something in compass-components, using across multiple packages is

File: packages/compass-connections/src/hooks/use-connection-status.ts
Comment (kmruiz) [reply packages/compass-connections/src/hooks/use-connection-status.ts:11]: > The reason to do so is that we don't want to make a requirement for some parts of the application to depend on the whole connection info object (that they generally don't care about), potentially causing unneccessary subscriptions to the state updates, heavy deep check to bypass referencial equality issues, and complex state propagation scenarios in cases where passing an id that uniquely identifies a connection should be enough.

That's not the case, and using objects as an argument to a function is not a problem. If you have a hook that has a dependency on the whole object in the dependency array (thanks React for that) yes, it can be problematic, but there is no subscription in this case anywhere. However, by passing the whole object, you gain the benefit of:

1. Encapsulation, if the object changes or the hook depends on other fields of ConnectionInfo, you just change the hook.
2. Intention is clear: you get the status of a connection, that is what you want to achieve, and the code clearly makes that explicit (doesn't care about the id).
3. We already have a reference to an object when we use useConnectionInfo (the only way to get the connection info), and in JavaScript mostly everything (unless some optimisations) is passed by reference. In this case, there is no additional impact.

If I'm missing something obvious I can change the code obviously, but I don't think it will generate additional subscriptions to anything by just using an object as an argument instead of a string.

Comment (gribnoysup) [reply packages/compass-connections/src/hooks/use-connection-status.ts]: > That's not the case, and using objects as an argument to a function is not a problem.

This is normal for functions, but you are not writing a function, you are writing a hook and hooks have a very spefic use-case: to be called inside React components.

> If you have a hook that has a dependency on the whole object in the dependency array (thanks React for that) yes, it can be problematic, but there is no subscription in this case anywhere

You are ignoring that to get that value to React rendering context to be able to pass it to the hook, it needs to pass through the props first (or get pulled from context), both cases mean that you are binding the component to state of a connection info

> Encapsulation, if the object changes or the hook depends on other fields of ConnectionInfo, you just change the hook

Realistically speaking, do we expect an id on the connecion info object not to be a sole identifier of a connection info? We should have a clear, documented way to identify a connection in the list of others. This is an id right now. Do we have any plans for connection info to be identified by something else? If yes, why would we design a system for ourselves that is poorly adapted to the other parts of the application code?

> We already have a reference to an object when we use useConnectionInfo (the only way to get the connection info), and in JavaScript mostly everything (unless some optimisations) is passed by reference. In this case, there is no additional impact.

Referential equality is currently preserved ONLY in the case of the old, single connection flow that this project will remove. Your multi connection code right now is written like this: connection info changes > code re-reads the list of connections from disk as the only existing way of updating the state > connection info object is completely overriden, you lost your reference (and due to the fact that the stores and hooks are all using local UI React state you have no referential equality between them). Same thing is implied by redux: changes in object require to create a new object reference, and state passed to the UI should be granular. As soon as we move this state to redux where it belongs, you will not have referential equality preserved even if we fix how the connectionInfo is updated. Relying on this is very brittle and will lead to hard-to-debug issues.

Comment (kmruiz) [reply packages/compass-connections/src/hooks/use-connection-status.ts]: Well the React magic on hooks arguments is something that I wasn't aware of and it's pretty unexpected. I appreciate the explanation because it's another design trap from the library.

I'll use the connectionInfoId as a parameter and resolve this comment.

Thanks a lot!

Comment (gribnoysup) [reply packages/compass-connections/src/hooks/use-connection-status.ts]: It's not about the hook function arguments, there is no magic there. To get the value to the hook you need to get it inside the render method from somewhere first. For example in this scenario currently it's the `useConnectionInfo` hook that is "dumb" enough in the current single connection flow implementation that the value set in the application root never changes after connection. But this will not hold up for the multiple connections scenario, it's already not holding up, the code that's currently written doesn't give any referential equality guarantees.

File: packages/compass-connections/src/hooks/use-connection-status.ts
Comment (gribnoysup) [inline packages/compass-connections/src/hooks/use-connection-status.ts:20]: This needs to be wrapped in useCallback, or you should move the definition inside effect. You are recreating this function every render unnecessarily adding and removing subscriptions from the connection manager instead of doing it once

File: packages/compass-connections/src/hooks/use-connection-status.ts
Comment (gribnoysup) [inline packages/compass-connections/src/hooks/use-connection-status.ts:32]: This doesn't currently have a dependency on the connection info id

**Review by gribnoysup — APPROVED** [review]
Please fix the hook, otherwise lgtm

File: packages/compass-connections/src/hooks/use-connection-status.ts
Comment (gribnoysup) [inline packages/compass-connections/src/hooks/use-connection-status.ts:32]: You are now missing dependencies here, it should include `connectionManager` and `connectionInfoId` in hooks deps.

**Pull request merged**
